diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -1845,3 +1895,49 @@ bool DifferentialFunctionComparator::isDependingOnReloc(
 
     return false;
 }
+
+/// Recursively collect operands of commutative binary operators.
+/// Leafs must be constants or already synchronized values.
+/// Return false if a non-synchronized non-constant leaf is found.
+bool DifferentialFunctionComparator::collectBinaryOperands(
+        const Value *Val,
+        Instruction::BinaryOps Opcode,
+        std::multiset<int> &SNs,
+        std::multiset<std::string> &Constants,
+        DenseMap<const Value *, int> &sn_map) const {
+    // Substitute an ignored value with its replacement when necessary
+    auto replacementIt = ignoredWithReplacement.find(Val);
+    const Value *newVal = replacementIt == ignoredWithReplacement.end()
+                                  ? Val
+                                  : replacementIt->second;
+    // Constant leaf
+    if (auto Const = dyn_cast<ConstantInt>(newVal)) {
+        Constants.insert(valueAsString(Const));
+        return true;
+    }
+    // Non-leaf; try to collect its operands recursively
+    auto BinOp = dyn_cast<BinaryOperator>(newVal);
+    if (BinOp && BinOp->getOpcode() == Opcode) {
+        std::multiset<int> tmpSNs;
+        std::multiset<std::string> tmpConstants;
+        if (collectBinaryOperands(
+                    BinOp->getOperand(0), Opcode, tmpSNs, tmpConstants, sn_map)
+            && collectBinaryOperands(BinOp->getOperand(1),
+                                     Opcode,
+                                     tmpSNs,
+                                     tmpConstants,
+                                     sn_map)) {
```

Comment:
Not very nice formatting, can we do something about it?

Reply:
Updated to use two variables to indicate success on each side of the operator.